### ISSUE TITLE
False positive when lifetime parameters are only used in `where` clauses

### ISSUE THREAD

**Issue by GearsDatapacks** [rust-lang/rust-clippy#13749]
### Summary

Clippy provides an incorrect suggestion when using lifetimes in `where` clauses.

### Lint Name

needless_lifetime

### Reproducer

I tried this code:

```rust
pub struct Generic<T>(T);

impl<'a, T> Generic<T> where T: 'a {}
```

I saw this happen:

```
warning: this lifetime isn't used in the impl
 --> src/lib.rs:3:6
  |
3 | impl<'a, T> Generic<T> where T: 'a {}
  |      ^^
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#extra_unused_lifetimes
  = note: `#[warn(clippy::extra_unused_lifetimes)]` on by default

warning: the following explicit lifetimes could be elided: 'a
 --> src/lib.rs:3:6
  |
3 | impl<'a, T> Generic<T> where T: 'a {}
  |      ^^                         ^^
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#needless_lifetimes
  = note: `#[warn(clippy::needless_lifetimes)]` on by default
help: elide the lifetimes
  |
3 - impl<'a, T> Generic<T> where T: 'a {}
3 + impl<T> Generic<T> where T: '_ {}
```

I expected to see this happen:
Clippy to not report a warning here, since lifetimes in `where` clauses can't be elided (to my knowledge)

### Version

```text
rustc 1.83.0 (90b35a623 2024-11-26)
binary: rustc
commit-hash: 90b35a6239c3d8bdabc530a6a0816f7ff89a0aaf
commit-date: 2024-11-26
host: x86_64-unknown-linux-gnu
release: 1.83.0
LLVM version: 19.1.
```

### Additional Labels

@rustbot label +I-suggestion-causes-error
